### PR TITLE
 prometheus-smartctl-exporter: unstable-2020-11-14 -> 0.9.0 

### PR DESCRIPTION
**Pull request by pshirshov**
prometheus-smartctl-exporter updated to 0.9.0, changelog; https://github.com/prometheus-community/smartctl_exporter/releases/tag/v0.9.0

Note: the yaml config is not supported anymore so I had to migrate options to cli arguments.

### CODE REVIEW

File: pkgs/servers/monitoring/prometheus/smartctl-exporter/default.nix [hunk 1]
```diff
@@ -1,36 +1,17 @@
-{ lib
-, fetchFromGitHub
-, fetchpatch
-, buildGoModule
-}:
+{ lib, fetchFromGitHub, fetchpatch, buildGoModule }:
```
Comment (mweinelt) [inline pkgs/servers/monitoring/prometheus/smartctl-exporter/default.nix:1]: Please don't reformat this.

File: pkgs/servers/monitoring/prometheus/smartctl-exporter/default.nix [hunk 2]
```diff
@@ -1,36 +1,17 @@
...
   pname = "smartctl_exporter";
-  version = "unstable-2020-11-14";
+  version = "0.9.0";
 
   src = fetchFromGitHub {
     owner = "prometheus-community";
     repo = pname;
-    rev = "e27581d56ad80340fb076d3ce22cef337ed76679";
-    sha256 = "sha256-iWaFDjVLBIAA9zGe0utbuvmEdA3R5lge0iCh3j2JfE8=";
+    rev = "v0.9.0";
```
Comment (mweinelt) [inline pkgs/servers/monitoring/prometheus/smartctl-exporter/default.nix:10]:
```suggestion
    rev = "refs/tags/v${version}";
```

We can reuse the version attribute here.

File: pkgs/servers/monitoring/prometheus/smartctl-exporter/default.nix [hunk 3]
```diff
@@ -1,36 +1,17 @@
...
+    sha256 = "sha256-z8z/A3BmaOSlS6/V0bEyVaUwV96pa7HM/0jad81394Q=";
```
Comment (mweinelt) [inline pkgs/servers/monitoring/prometheus/smartctl-exporter/default.nix:11]:
```suggestion
    hash = "sha256-z8z/A3BmaOSlS6/V0bEyVaUwV96pa7HM/0jad81394Q=";
```

I prefer using hash these days.

**Review by mweinelt — CHANGES_REQUESTED** [review]

**Comment by mweinelt** [comment]
The commit message should say `prometheus-smartctl-exporter: unstable-.... -> 0.9.0`. 

Please do not delete the PR template and make sure you read CONTRIBUTING.md.

**Comment by Artturin** [comment]
@pshirshov iirc you have to be reminded about the commit message format nearly every PR, please try to remember it. Maybe make a checklist for yourself?

**Comment by pshirshov** [comment]
Ok, now it works. This update solves numerous issues of the old version.

**Comment by pshirshov** [comment]
@mweinelt : could you merge it now?

**Comment by mweinelt** [comment]
Package and module changes should go into separate commits.

File: nixos/modules/services/monitoring/prometheus/exporters/smartctl.nix [hunk 4]
```diff
@@ -4,16 +4,7 @@ with lib;
 
 let
   cfg = config.services.prometheus.exporters.smartctl;
-  format = pkgs.formats.yaml {};
-  configFile = format.generate "smartctl-exporter.yml" {
-    smartctl_exporter = {
-      bind_to = "${cfg.listenAddress}:${toString cfg.port}";
-      url_path = "/metrics";
-      smartctl_location = "${pkgs.smartmontools}/bin/smartctl";
-      collect_not_more_than_period = cfg.maxInterval;
```
Comment (mweinelt) [inline nixos/modules/services/monitoring/prometheus/exporters/smartctl.nix:13]: Looks like you removed this option. Use `mkRemovedOptionModule`.

Comment (pshirshov) [reply nixos/modules/services/monitoring/prometheus/exporters/smartctl.nix:13]: That's not an option, that's the content of the config file which I migrated to cli arguments. There is no config file anymore.

Comment (mweinelt) [reply nixos/modules/services/monitoring/prometheus/exporters/smartctl.nix:13]: I missed that you migrated that option, so I resolved this review.

**Comment by pshirshov** [comment]
> make sure you read CONTRIBUTING.md.

It's complicated, I can't remember that because I'm just making occasional patches in my spare time.

**Comment by mweinelt** [comment]
https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes

**Comment by pshirshov** [comment]
Well, something is definitely broken if I have to memoize that there must be no period at the end of commit message.

The link is not helpful. I don't know what is wrong. I don't know what is a nixos module, maybe there was a change in the module and I have to do some rituals to reflect thah or maybe I forgot to add motivation to my changes but I believe that version bump itself is enough of motivation.

The experience of submitting a patch to nix is abysmal exactly because of these situations, people don't want to be bullied because of odd dot at the end of a commit message. Less they want to infer what exactly is wrong from a document written by an autist for another autist. Though I'm an autist and still cannot figure out what is wrong with my commit messages.

It's indeed easier to keep my patches in my own repo. I understand that it's a great honour when you accept a patch into your main tree but seemingly I don't deserve that honour.

**Comment by SuperSandro2000** [comment]
> The experience of submitting a patch to nix is abysmal

No, it isn't, you are just unfamiliar with it and ranting about it by now. It is definitely far from perfect and has lots of pieces that can be improved but compared to other linux distros it is more accessible. 

> I don't know what is a nixos module

taken from https://nixos.wiki/wiki/NixOS_modules which was the first google result:

```
Modules are files combined by NixOS to produce the full system configuration. A module contains a Nix expression. It declares options for other modules to define (give a value). It processes them and defines options declared in other modules.[[1]](https://nixos.wiki/wiki/NixOS_modules#cite_note-1)

For example, /etc/nixos/configuration.nix is a module. Most other modules are in Nixpkgs-link.png﻿[nixos/modules](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules).
```

This means the change in ``nixos/modules/services/monitoring/prometheus/exporters/smartctl.nix`` is the module change and the commit message should be something like ```
nixos/prometheus-smartctl: change to cli options

config file support was removed in 0.8.0
```

**Comment by pshirshov** [comment]
Thanks for the explanation. Does it look better now?..

Anyway, it's still not obvious what is going to be considered "correct".

Re other distros. I had limited experience with sending patches to arch/gentoo and many times it was just "thanks for the patch, we will polish the rough edges", not minor formalities. But anyway, I've been maintaning my own overlay for years.

**Comment by mweinelt** [comment]
It is often helpful to just look at the git log for the file you're editing to glean some info on how the commit messages are structured.

**Comment by pshirshov** [comment]
So, is it better not, can it be merged?..

**Comment by Artturin** [comment]
the commit messages were flipped so i fixed them

**Comment by mweinelt** [comment]
With my previous config it crashes. It seems to ignore the devices I pass.

```
ExecStart=/nix/store/w4mf1lfx9x5q8xgwz6mvabrn56pbgn5x-smartctl_exporter-0.9.0/bin/smartctl_exporter --web.listen-address="0.0.0.0:9633" --smartctl.path="/nix/store/picpj8rnjnsdqif9y4zakz9glknamhsw-smartmontools-7.3/bin/smartctl" --smartctl.interval="60s" --smartctl.device=/dev/nvme0n1 --smartctl.device=/dev/nvme1n1
```

```
systemd[1]: Started prometheus-smartctl-exporter.service.
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.146Z caller=main.go:94 level=info msg="Starting smartctl_exporter" version="(version=, branch=, revision=)"
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.146Z caller=main.go:95 level=info msg="Build context" build_context="(go=go1.19.3, user=, date=)"
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.158Z caller=main.go:116 level=warn msg="Device unavailable" name=/dev/nvme0n1
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.158Z caller=main.go:116 level=warn msg="Device unavailable" name=/dev/nvme1n1
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.158Z caller=main.go:123 level=info msg="No devices specified, trying to load them automatically"
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.168Z caller=readjson.go:68 level=warn msg="S.M.A.R.T. output reading" err="exit status 2"
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.168Z caller=readjson.go:118 level=error msg="Device open failed, device did not return an IDENTIFY DEVICE structure, or device is in a low-power mode"
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.168Z caller=readjson.go:150 level=error msg="Smartctl open device: /dev/nvme0 failed: Permission denied"
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.168Z caller=main.go:57 level=error msg="Error collecting SMART data" err="smartctl returned bad data for device /dev/nvme0"
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.178Z caller=readjson.go:68 level=warn msg="S.M.A.R.T. output reading" err="exit status 2"
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.178Z caller=readjson.go:118 level=error msg="Device open failed, device did not return an IDENTIFY DEVICE structure, or device is in a low-power mode"
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.178Z caller=readjson.go:150 level=error msg="Smartctl open device: /dev/nvme1 failed: Permission denied"
smartctl_exporter[1809638]: ts=2022-11-07T21:44:56.178Z caller=main.go:57 level=error msg="Error collecting SMART data" err="smartctl returned bad data for device /dev/nvme1"
smartctl_exporter[1809638]: panic: runtime error: index out of range [0] with length 0
smartctl_exporter[1809638]: goroutine 43 [running]:
smartctl_exporter[1809638]: main.(*SMARTctlInfo).mineVersion(0xc0000a9e08)
smartctl_exporter[1809638]:         github.com/prometheus-community/smartctl_exporter/smartctlinfo.go:59 +0xba5
smartctl_exporter[1809638]: main.(*SMARTctlInfo).Collect(...)
smartctl_exporter[1809638]:         github.com/prometheus-community/smartctl_exporter/smartctlinfo.go:48
smartctl_exporter[1809638]: main.SMARTctlManagerCollector.Collect({{0x0, 0x0}, 0x0, {0xc0002087a0, 0x2, 0x2}, {0xa788a0, 0xc00023c300}}, 0xc000254120?)
smartctl_exporter[1809638]:         github.com/prometheus-community/smartctl_exporter/main.go:60 +0xd2
smartctl_exporter[1809638]: github.com/prometheus/client_golang/prometheus.DescribeByCollect.func1()
smartctl_exporter[1809638]:         github.com/prometheus/client_golang@v1.13.0/prometheus/collector.go:90 +0x2b
smartctl_exporter[1809638]: created by github.com/prometheus/client_golang/prometheus.DescribeByCollect
smartctl_exporter[1809638]:         github.com/prometheus/client_golang@v1.13.0/prometheus/collector.go:89 +0x9c
systemd[1]: prometheus-smartctl-exporter.service: Main process exited, code=exited, status=2/INVALIDARGUMENT

**Comment by mweinelt** [comment]
The hardening is not at fault.

```
# systemd-run -p DynamicUser=true -p SupplementaryGroups=disk -p AmbientCapabilities=CAP_SYS_ADMIN -t  smartctl --all /dev/nvme0n1
Running as unit: run-u53835.service
Press ^] three times within 1s to disconnect TTY.
smartctl 7.3 2022-02-28 r5338 [x86_64-linux-5.19.17] (local build)
Copyright (C) 2002-22, Bruce Allen, Christian Franke, www.smartmontools.org

=== START OF INFORMATION SECTION ===
Model Number:                       Samsung 980 PRO with Heatsink 2TB

**Comment by pshirshov** [comment]
This snippet may/should fix your issue:

```
  services.prometheus.exporters = {
    smartctl = {
      group = "disk";
    };
  };

  services.udev.extraRules = ''
    SUBSYSTEM=="nvme", GROUP="disk"
    SUBSYSTEM=="block", GROUP="disk"
  '';
```

If it does the problem is not relevant to/caused by the version update.

**Pull request closed without merging**